lightning/src/ln/peers/handshake/states.rs

```diff
@@ -1,30 +1,888 @@
-use ln::peers::handshake::hash::HandshakeHash;
+use bitcoin::secp256k1;
+
+use bitcoin::hashes::{Hash, HashEngine};
+use bitcoin::hashes::sha256::Hash as Sha256;
 use bitcoin::secp256k1::{SecretKey, PublicKey};
 
-pub enum HandshakeState {
-	Uninitiated,
-	AwaitingActOne(ActOneExpectation),
-	AwaitingActTwo(ActTwoExpectation),
-	AwaitingActThree(ActThreeExpectation),
-	Complete,
+use ln::peers::{chacha, hkdf5869rfc};
+use ln::peers::conduit::{Conduit, SymmetricKey};
+use ln::peers::handshake::acts::{Act, ActBuilder, ACT_ONE_LENGTH, ACT_TWO_LENGTH, ACT_THREE_LENGTH, EMPTY_ACT_ONE, EMPTY_ACT_TWO, EMPTY_ACT_THREE};
+use ln::peers::handshake::IHandshakeState;
+
+// Alias type to help differentiate between temporary key and chaining key when passing bytes around
+type ChainingKey = [u8; 32];
+
+// Generate a SHA-256 hash from one or more elements concatenated together
+macro_rules! concat_then_sha256 {
```

Comment:
I still think this macro might be more legibly put in a function, or a custom hash object/class. But, as I said before, ultimately this is up to Matt and Antoine.

Reply:
@ariard @arik-so

This would have been better in the right PR, but I appreciate the persistence if you aren't quite sold.

Overall, I think it makes the code less readable and introduces a wrapper type that has to be unwrapped which I think detracts from the audit-ability. I'm also not sure about the aversion to variadic macros. I needed a way to do the hash for 1 or 2 objects and used a macro that made the code (in my point of view) more legible.

With that said, I did implement it so the two of you can look and see which you prefer. Just let me know and I can fixup (1/4) to have the final result.

https://github.com/julianknutsen/rust-lightning/commit/0b4c819d83d2d3fef43ad0afd27c5431228f9fbb

Reply:
I moved this to 1/4 on rust-lightning PR

Reply:
Thanks to implement the alternative that's useful to have a look.

Given the compactness of the macro and the really clear callsites where it's used I'm leaning to keep the macro. That said I share the more general feeling of Arik wrt to how we use macro in the rest of the codebase, which is sadly hurting code readability....